### DISCUSSION TITLE
Simple Worship Scratch Pad

### DISCUSSION THREAD

**Discussion by applehat** [solafide-dev/gobible#1] [Ideas]
Initial base software ideas.

# Main Mission
I think the main missional idea for this project comes from The love for small churches and church plants, faithfully serving the lord with limited resources
- We want software that can run on an old laptop and a cheap projector, but is still useful through the life of your church if the Lord chooses to grow you to hundreds or thousands of people
- We want it to be free. I'd rather see faithful Christians be the hands and feet with their money then pay for a subscription service.

Secondarily this this tho, we are nerds. And we know many other people in the tech side of ministry are too. If I want to trigger slide changes based off a custom built esp32 button, I want software that I can interface with reliable and falt-tollerantly, not some hacky work around that I worry will fail.

# Current Core Wants
- Robust, pluggable, events based system
- Use Web Technologies (Wails + React, maybe?)
- Server displays using HTTP so that remote (or even main) displays can be easily consumed by cheap commodity hardware
- Slides with text / photos that are simple to use but still powerful for the advanced user
- Bible integrations
- Song library
- Ideally an easy to follow order of service view -- a huge screen of slides can be fairly distracting and hard to follow for folks trying to run presentations
- A quick access "blank" screen that is customizable. I find that running slides I often want to just display the church logo or something similar when we don't have anything else relevant to display on screen.

## Song Library
- Songs should be defined in a single file using something like markdown, and then song orders can just be a CSV of song parts. 
- Slide generation should happen automatically based off this data (making slides for songs is a pain and no one enjoys that)

## Bible Integration
- Use the gobible library -- Potentially expanding and making it better in the process

## Ideas for Plugins

These are all potential plugin ideas once a simple base software exists. I'm writing this here not to cause feature creep in the main code base, but to think of everything we might want to do in the future that we should consider when designing the base software.

- REST API
- Remote Control (App, Buttons, IR, or even a dumb alexa integration if someone wants to write the plugin)
- Church Trac / Planning Center / Whatever else integration for automated Order of Service creation
- Event Triggered MIDI (or MIDI input to event)
- Catechism / Confession text importers 
- Online song library integrations 
- A Webhook / Local Script trigger plugin
- Sermon Audio auto recording

**Comment by applehat** [comment]
So I was looking at Wails (vs something like Electron) and I think it still fits the bill, but I will note that it currently doesn't seem to offer "multiple window" support.

I think this is fine, if we follow the "the display is HTTP". 

If SimpleWorship runs as a "server and interface" -- then we can write many additional display apps for various devices.
- Wails based local display app for Windows / OSX / Linux (raspberry pi image to auto-launch and connect even?)
- iOS
- Android
  - Google TV
  - Fire TV
  

We can use the SSDP protocol to broadcast the servers alive-ness constantly, and the clients can just find and connect to it with no configuration needed for simplicity from the user.
  
 https://github.com/koron/go-ssdp
  
Alternatively it might make sense for the display app to be written in React Native -- For a single app that can work across all the above use cases -- I just have little experience with React Native, and I'm a giant golang fanboy -- so I'm biased. Hah

In fact, there's a react native ssdp package as well that would probably make it a fairly easy project to tie together. https://www.npmjs.com/package/react-native-ssdp

**Reply by michaelpanik** [reply]
I’m all about it. I think wails and react is a good stack as well. Other plugin ideas:
- blackmagic capture cards
- Video backgrounds
- Propresenter importer 😏

Are videos good core functionality? A lot of churches, even small ones, will use movie clips and things for sermon illustrations, plus bumper videos, etc. Maybe plugins for hosted videos like YouTube and Vimeo, but local video files as a core function

**Reply by applehat** [reply]
In my mind the "displays" are just webviews, so they can serve whatever content.

This is actually a place where I struggled when working on this idea in the past tho -- I got to the point where I had a local database of slides and struggled with how to render them.

Something I'd also add as a core feature: portability.

I'd like this to be so "church plant" usable that it could run off a thumb drive potentially. My church plant shares space with another church and we meet in the afternoon -- being able to store our app on a thumb drive and not even have to install anything on their machine would be amazing.

**Reply by michaelpanik** [reply]
That shouldn’t be a problem. Most applications can run off a thumb drive. I would say that’s a vote for making it cloud based though, but that incurs some overhead in cost. I don’t mind covering that cost, just a thought. The online data should be offline and synced constantly though. But like you could open the app, log in, and there’s all your stuff. If you’re offline, that data has been synced last time you were online and it’s available. If you’ve never been online, you can load from a local file.

**Reply by applehat** [reply]
I say you just call that cloud syncing, right?

And it could be done totally free if folks just linked something like a google drive or drop box account and the file storage is done with static files vs a database

**Reply by michaelpanik** [reply]
It actually could be a SQLite file even if you want a db, but yeah a bring your own cloud solution is a great idea.

**Reply by applehat** [reply]
I think one reason I like the idea of a "flat file" type storage -- even though its not necessarily the highest performance option, I think it makes certain "power user" tasks even easier.

You mentioned a pro-presenter importer -- It'd be great if that kind of tool didn't even really have to interface with SW in any real way -- and instead just generated the flat files SW uses. (I just like human readable and quickly hackable things -- especially when away from my dev environment)

**Reply by michaelpanik** [reply]
Yeah sorry - not sure why I even mentioned the whole DB thing again. What I've built so far was all written around JSON representations of the slideshows, and playlists JSON files that link them. Makes more sense almost definitely.
